h2o-algos/src/main/java/hex/tree/TreeHandler.java

```diff
@@ -0,0 +1,195 @@
+        final String trimmedCategorical = categorical.trim(); // Trim the categorical once - input from the user
+
+        switch (sharedTreeOutput.getModelCategory()) {
+            case Binomial:
+                // Fall through to regression - handling is equal
+            case Regression:
+                return 0; // There is only one tree for regression and binomial
+            default:
+                // Search for the index of categorical in case of multinomial
+                for (int i = 0; i < responseColumnDomain.length; i++) {
+                    // User is supposed to enter the name of the categorical level correctly, not ignoring case
+                    if (trimmedCategorical.equals(responseColumnDomain[i])) return i;
+                }
+                break;
+        }
+        // Not a regression or binomial and the given categorical has not been found in the response column's domain.
+        return -1;
+    }
+
+    /**
+     * @param args An instance of {@link TreeV3} input arguments to validate
+     * @param output An instance of {@link SharedTreeModel.SharedTreeOutput} to validate input arguments against
+     * @param responseLevelIndex
+     */
+    private static void validateArgs(TreeV3 args, SharedTreeModel.SharedTreeOutput output, final int responseLevelIndex) {
+        if (args.tree_number < 0) throw new IllegalArgumentException("Tree number must be greater than 0.");
+
+        if (args.tree_number > output._treeKeys.length - 1)
+            throw new IllegalArgumentException("There is no such tree.");
+
+        if (responseLevelIndex < 0)
+            throw new IllegalArgumentException("There is no such tree class. Given categorical level does not exist in response column: " + args.tree_class.trim());
+    }
+
+
+    /**
+     * Converts H2O-3's internal representation of a boosted tree in a form of {@link SharedTreeSubgraph} to a format
+     * expected by H2O clients.
+     *
+     * @param sharedTreeSubgraph An instance of {@link SharedTreeSubgraph} to convert
+     * @return An instance of {@link TreeProperties} with some attributes possibly empty if suitable. Never null.
+     */
+    static TreeProperties convertSharedTreeSubgraph(final SharedTreeSubgraph sharedTreeSubgraph) {
+        Objects.requireNonNull(sharedTreeSubgraph);
+
+        final TreeProperties treeprops = new TreeProperties();
+
+        treeprops._leftChildren = MemoryManager.malloc4(sharedTreeSubgraph.nodesArray.size());
+        treeprops._rightChildren = MemoryManager.malloc4(sharedTreeSubgraph.nodesArray.size());
+        treeprops._descriptions = new String[sharedTreeSubgraph.nodesArray.size()];
+        treeprops._thresholds = MemoryManager.malloc4f(sharedTreeSubgraph.nodesArray.size());
+        treeprops.features = new String[sharedTreeSubgraph.nodesArray.size()];
+        treeprops.nas = MemoryManager.mallocZ(sharedTreeSubgraph.nodesArray.size());
+
```

Comment:
The strings may use much more memory and are in fact expected to use it. No matter the size of the tree (hundreds of nodes max usually), these strings are expected to be more memory intensive.

Yet there is no reaason not to tell our MemoryManager about memory usage if we can. Costs us nothing. Even if the memory is allocated for really short time, like in this case.

Reply:
looks like using a sledgehammer to crack a nut, imo.
MM seems to have been designed for persistent serializable data (vecs, chunks and so on), it should also ensure that there's always memory available for usual small operations; basically if you run OoM when creating a short-lived array being returned to client, it means that the memory issue is somewhere else and it was about to run OoM anyway... that's why it looks like premature optimization.

Reply:
The point of using a sledgehammer to crack a nut is simple - there is greater cost of using a sledgehammer instead of a normal hammer, or even a fist or your head. Here, the sledgehammer is absolutely for free (maybe one more method invocation in theory, in reality after inlining, the cost would be +- the same).

Using it costs nothing more. Not using it will not bring a running instance of H2O any benefits most of the time, but the more allocations are known to the MemoryManager, the better it's performance is.